### DISCUSSION TITLE
How does the cross attention work?

### DISCUSSION THREAD

**Discussion by SinanAkkoyun** [openai/whisper#943] [Q&A] [answered]
Hello,
I am trying to understand the autoregressiveness of the decoder.
Regarding the audio feature cross-attention: Does it for every iteration of the for loop (decoder.py, line 595 _main_loop) take ALL audio features (xa) simultaneously for each cross attention decoding or does it only take specific samples?

I want to understand how parallel/serial the whole approach is. I highly appreciate any answer!

**Comment by SinanAkkoyun** [comment, marked as answer]
@jongwook I hope the ping is ok

Forget what I asked above, this is the right question:
Given the for loop is at position 2s of a 5s wav file, it takes all 5s audio_features for prediction at any point, right? So, does predicting the token at timepoint 2s also take future audio_features into consideration?
What would happen if the decoder only had access to the current audio_features?

**Reply by jongwook** [reply]
You are right. With the current architecture, the prediction is based on the entire audio in the 30-second window. This allows the prediction to more flexibly refer to the audio features out-of-order, say if the audio said "thirty million dollars", Whisper tends to output "$30 million", putting the dollar sign before "thirty million" is spoken. This capability is also useful for translation, where the word ordering is often different between languages.

**Reply by SinanAkkoyun** [reply]
Thank you so much for your answer. I am trying to implement true live asr streaming (not running the whole inference per sentence but true live transcription for very fast speeds), thus I am interested in adding small audio chunks one by one.
In my private testing I noticed the following (I made the _main_loop use a newly sliced audio file each iteration): When I set the audio lenght for the first iteration to zero or something very small, it only generates until that point. This happens despite me giving the model the full audio clip after i < 1.
This leads me to believe, that the cross attention from the get go, i = 0 blocks off the right paths of the decoder, thus no matter if I suppy everything later, it won't work. Is that true? (either that or I did something wrong)

No matter that detail, I would really love to do true live stream transcription, leveraging for example kv cache of the already transcribed audio, but the cross attention of full 30s is in the way if I understand correctly.

Do you think there is any solution to this? Thank you again!

**Reply by atyshka** [reply]
I would also be interested in something similar... the current approaches that are running "live" with 30s chunks aren't good enough for true real time on-device inference. Google ASR has a really nice live transcription option and it would be great if we could do that with whisper as well.
I think the whisper model would definitely require modifications, mainly masked attention rather than full attention, as well as relative positional embeddings. Let me know if anyone else plans to work on this, I'd love to contribute to either the model design or an implementation in whisper.cpp

**Reply by SinanAkkoyun** [reply]
Awesome, after we hopefully gather some more intel we could work on that, I've been working on it and an interface for some time, planning on PRing when it's done

Yes, masked attention will be necessary, but also will require retraining... Regarding the relative positional embedding, why would that be needed? The sin pos encoding does scale from the beginning of a sentence, no?

**Reply by atyshka** [reply]
I guess the question is how would you handle sequences longer than 30 seconds? I’d envision chopping off the previous audio so you have a constant window of 30s audio. But if you do that, the sinusoidal position embeddings no longer start at 0… I’m not sure if that would be a problem or not

**Reply by SinanAkkoyun** [reply]
That's no problem. No single sentence is more than 30 seconds long. My plan is to do VAD to detection and reset the positional embeddings for every sentence. Contextual robustness of that level (inter sentences context is negligible for real time detection)

**Reply by SinanAkkoyun** [reply]
The bigger problem lies within the architecture and retraining... The whole decoder cross-attention needs to be replaced and the decoder retrained...

**Reply by atyshka** [reply]
The encoder also needs to be retrained, right? But yeah, the bigger problem is there's no code provided for training from scratch, and I'm not sure how many GPU-hours a model like this would take to train.

**Comment by SinanAkkoyun** [comment]
@jongwook Are there any resources on how to retrain the decoder? (instead of fine tuning it) Or do you have an idea on how to solve the above without major retraining?

**Reply by jongwook** [reply]
The decoder is autoregressive while the encoder is not. Encoder needs only one forward pass, so running the encoder on every slice using a sliding 30s window (with ~1s stride) is not a big performance hit, relatively. More disciplined way is use a Transducer model like [this](https://huggingface.co/nvidia/stt_en_conformer_transducer_xlarge), but it'd be an entirely different architecture and probably requires training from scratch.

**Reply by SinanAkkoyun** [reply]
I see, thank you very much. You mean different architecture than whisper or than the nvidia model?

**Reply by jongwook** [reply]
The linked nvidia model is an example of transducer models, which works differently (while still being transformer-based) from Whisper, which is an encoder-decoder model. [This blog article](https://www.assemblyai.com/blog/an-overview-of-transducer-models-for-asr/) is a great introduction on how transducer models work.
